compiler/rustc_session/src/errors.rs

```diff
@@ -4,23 +4,51 @@ use rustc_ast::token;
 use rustc_ast::util::literal::LitError;
 use rustc_errors::{
     codes::*, Diag, DiagCtxt, DiagMessage, Diagnostic, EmissionGuarantee, ErrorGuaranteed, Level,
-    MultiSpan,
+    MultiSpan, Subdiagnostic,
 };
 use rustc_macros::Diagnostic;
 use rustc_span::{Span, Symbol};
 use rustc_target::spec::{SplitDebuginfo, StackProtector, TargetTriple};
 
 use crate::{config::CrateType, parse::ParseSess};
 
+// FIXME: factor into separate structs to avoid dynamic DiagMessage field
 pub(crate) struct FeatureGateError {
     pub(crate) span: MultiSpan,
     pub(crate) explain: DiagMessage,
+    pub(crate) subdiag: FeatureGateSubdiagnostic,
 }
 
 impl<'a, G: EmissionGuarantee> Diagnostic<'a, G> for FeatureGateError {
     #[track_caller]
     fn into_diag(self, dcx: &'a DiagCtxt, level: Level) -> Diag<'a, G> {
-        Diag::new(dcx, level, self.explain).with_span(self.span).with_code(E0658)
+        let mut diag = Diag::new(dcx, level, self.explain).with_span(self.span).with_code(E0658);
+        diag.subdiagnostic(dcx, self.subdiag);
+        diag
+    }
+}
+
+pub struct FeatureGateSubdiagnostic {
+    pub(crate) issue: Option<FeatureDiagnosticForIssue>,
+    pub(crate) enable_feature: Option<EnableFeatureSubdiagnostic>,
+    pub(crate) upgrade_compiler: Option<SuggestUpgradeCompiler>,
+}
+
+impl Subdiagnostic for FeatureGateSubdiagnostic {
+    fn add_to_diag_with<G: EmissionGuarantee, F: rustc_errors::SubdiagMessageOp<G>>(
+        self,
+        diag: &mut Diag<'_, G>,
+        f: F,
+    ) {
+        if let Some(issue) = self.issue {
+            issue.add_to_diag_with(diag, |diag, msg| f(diag, msg));
```

Comment:
I'm not happy about this, ideally `F` would be a reference so it can be copied into nested `add_to_diag_with()` calls instead of being wrappen in another closure.